src/xia2/cli/cluster_analysis.py

```diff
@@ -0,0 +1,226 @@
+    .short_caption = "SSX flag"
+
+output {
+  log = xia2.multiplex.log
+    .type = str
+}
+""",
+    process_includes=True,
+)
+
+mca_phil = iotbx.phil.parse(
+    """
+include scope xia2.cli.report.phil_scope
+
+include scope xia2.Modules.MultiCrystal.ScaleAndMerge.phil_scope
+
+seed = 42
+  .type = int(value_min=0)
+
+unit_cell_clustering {
+  threshold = 5000
+    .type = float(value_min=0)
+    .help = 'Threshold value for the clustering'
+  log = False
+    .type = bool
+    .help = 'Display the dendrogram with a log scale'
+}
+
+ssx_flag = False
+    .type = bool
+    .help = "Set this to true for analysing clusters of ssx_data"
+    .short_caption = "SSX flag"
+
+output {
+  log = xia2.multi_crystal_analysis.log
+    .type = str
+}
+""",
+    process_includes=True,
+)
+
+
+def run(args=sys.argv[1:]):
+    # Create the parser
+
+    usage = "xia2.cluster_analysis [options] [param.phil] scaled.expt scaled.refl"
+
+    help_message = """
+    Will run the intensity and cos angle clustering methods on a multi-crystal dataset.
+    """
+    parser = ArgumentParser(
+        usage=usage,
+        phil=phil_scope,
+        read_reflections=True,
+        read_experiments=True,
+        check_format=False,
+        epilog=help_message,
+    )
+
+    # Parse the command line
+    params, options = parser.parse_args(args=args, show_diff_phil=False)
+    params_mca = mca_phil.extract()
+
+    params_mca.ssx_flag = params.ssx_flag
+
+    if len(params.input.experiments) == 0:
+        print("No Experiments found in the input")
+        parser.print_help()
+        return
+    if len(params.input.reflections) == 0:
+        print("No reflection data found in the input")
+        parser.print_help()
+        return
+    try:
+        assert len(params.input.reflections) == len(params.input.experiments)
+    except AssertionError:
+        raise sys.exit(
+            "The number of input reflections files does not match the "
+            "number of input experiments"
+        )
+
+    if params.seed is not None:
+        flex.set_random_seed(params.seed)
+        np.random.seed(params.seed)
+        random.seed(params.seed)
+
+    experiments = flatten_experiments(params.input.experiments)
+    reflections = flatten_reflections(params.input.reflections)
+    if len(experiments) < 2:
+        sys.exit("xia2.cluster_analysis requires a minimum of two experiments")
+    reflections = parse_multiple_datasets(reflections)
+    experiments, reflections = assign_unique_identifiers(experiments, reflections)
+
+    reflections_all = flex.reflection_table()
+    assert len(reflections) == 1 or len(reflections) == len(experiments)
+    for i, (expt, refl) in enumerate(zip(experiments, reflections)):
+        reflections_all.extend(refl)
+    reflections_all.assert_experiment_identifiers_are_consistent(experiments)
+
+    try:
+        MCA = MultiCrystalReport(params_mca, experiments, reflections_all)
+
```

Comment:
Very excited to see this functionality coming to xia2 😃 . I'm just starting to test now, and I got the following error at this line:
```  
File "/dials/modules/xia2/src/xia2/cli/cluster_analysis.py", line 141, in run
    MCA._cluster_analysis.cc_clusters
AttributeError: 'NoneType' object has no attribute 'cc_clusters'
```
I think the clustering was not being run, is this line missing?
```suggestion
MCA.cluster_analysis()
```
Also, in order to run the program, it looks like you need to supply a phil file with at least the parameters `max_cluster_height_difference`,`min_cluster_size`, `max_output_clusters`. The standard way we do this in the programs is to have all options defined in the phil scope with suitable default values and a nice description, so that it is not necessary to provide a specific phil file as input. Then, if a phil file is provided, this is a mechanism to overwrite the default values.

Reply:
Yep! Unsure how the missing line happened - but I've put it back in :) 

Also confused myself with pushing separate branches and forgot to add the phil params - I put them in ScaleAndMerge.py since that's where most of the multiplex parameters are and this particular function of finding interesting clusters could eventually find use in multiplex as well - should be in this thread now :)